_posts/2022-01-10-reassign-partitions.md

```diff
@@ -0,0 +1,363 @@
+
+2. With the help of this tool, you can increase the no. of partitions / replicas which can help with increasing the throughput of the topic.
+
+## Example time
+
+Let us understand the working of this tool with an interesting problem
+Suppose we have 5 Kafka Brokers and after looking at the partition details we get to realize that the brokers are not too busy, so we can scale them down to 3.
+Through this example we will take a look at how the three phases of the Kafka reassignment partition tool(`--generate`, `--execute` and `--verify`) works.
+We will generate the JSON data that will be used in the `reassignment.json` file.
+We will then assign the partitions to the remaining broker using the `reassignment.json` file.
+The Kafka Cluster that we will use, will be configured to use PLAIN listener.
+Strimzi supports *TLS*, *SCRAM-SHA-512*, *OAUTH*, and *PLAIN* configuration options for authentication.
+
+Before proceeding towards the steps. Let's discuss one more curious question. Can you scale down any pod you want through this process?
+So the answer to this question is no. Wondering why?
+It is due to the fact Strimzi uses StatefulSets to manage broker pods. So you cannot remove any pod from the cluster.
+You can only remove one or more of the highest numbered pods from the cluster. For example, in a cluster of 5 brokers the pods are named `<CLUSTER-NAME>-kafka-0` up to `<CLUSTER-NAME>-kafka-4`.
+If you decide to scale down by two brokers, then `<CLUSTER-NAME>-kafka-4` and `<CLUSTER-NAME>-kafka-3` will be removed.
+
+### Preparing to scale down the number of Kafka Brokers
+
+This example will use a Kafka cluster deployed with the Strimzi Cluster Operator.
+Before we start with anything, we have to install the Strimzi Cluster Operator and deploy the Kafka cluster.
+You can install the Cluster Operator with any installation method you prefer.
+The Kafka cluster is then deployed with the PLAIN listener enabled on port 9092.
+
+Example Kafka configuration with PLAIN authentication.
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: <CLUSTER-NAME>
+spec:
+  kafka:
+    replicas: 5
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+    storage:
+      type: jbod
+      volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+  zookeeper:
+    replicas: 3
+    storage:
+      type: persistent-claim
+      size: 100Gi
+  entityOperator:
+    topicOperator: {}
+    userOperator: {}
+```
+
+Once the cluster is running, lets deploy some topics where we will send and receive the messages.
+Here is an example topic configuration. You can change it to suit your requirements.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaTopic
+metadata:
+  name: my-topic-1
```

Comment:
This is a subtle point and I'm not asking you to change it... topics with names like `{name}-{number}` can result in confusing broker logs. The broker logs _partitions_ using this same pattern, `{topicname}-{number}`. i.e. the second partition of this topic would be logged as `my-topic-1-1`. It creates an ambiguity between topic names and partitions. For this reason I personally would avoid naming a topic to end with a number, so that I could more easily `grep` the broker logs for `my-topic-1` and know I was getting hits about a partition, not a topic.